relayer/crates/starknet-chain-components/src/impls/queries/token_address/create.rs

```diff
@@ -0,0 +1,68 @@
+use core::marker::PhantomData;
+
+use cgp::prelude::*;
+use hermes_cairo_encoding_components::strategy::ViaCairo;
+use hermes_cairo_encoding_components::types::as_felt::AsFelt;
+use hermes_chain_components::traits::send_message::CanSendSingleMessage;
+use hermes_chain_type_components::traits::types::address::HasAddressType;
+use hermes_chain_type_components::traits::types::denom::HasDenomType;
+use hermes_encoding_components::traits::decode::CanDecode;
+use hermes_encoding_components::traits::encode::CanEncode;
+use hermes_encoding_components::traits::has_encoding::HasEncoding;
+use hermes_encoding_components::traits::types::encoded::HasEncodedType;
+use starknet::core::types::{Call, Felt};
+use starknet::macros::selector;
+
+use crate::impls::types::address::StarknetAddress;
+use crate::impls::types::message::StarknetMessage;
+use crate::traits::queries::contract_address::CanQueryContractAddress;
+use crate::traits::queries::token_address::{
+    CosmosTokenAddressOnStarknetQuerier, CosmosTokenAddressOnStarknetQuerierComponent,
+};
+use crate::types::message_response::StarknetMessageResponse;
+use crate::types::messages::ibc::denom::PrefixedDenom;
+
+#[cgp_new_provider(CosmosTokenAddressOnStarknetQuerierComponent)]
+impl<Chain, Encoding> CosmosTokenAddressOnStarknetQuerier<Chain>
+    for CreateCosmosTokenAddressOnStarknet
+where
+    Chain: HasAsyncErrorType
+        + HasEncoding<AsFelt, Encoding = Encoding>
+        + HasAddressType<Address = StarknetAddress>
+        + HasDenomType<Denom = StarknetAddress>
+        + CanSendSingleMessage<Message = StarknetMessage, MessageResponse = StarknetMessageResponse>
+        + CanQueryContractAddress<symbol!("ibc_ics20_contract_address")>
+        + CanRaiseAsyncError<Encoding::Error>,
+    Encoding: HasEncodedType<Encoded = Vec<Felt>>
+        + CanEncode<ViaCairo, PrefixedDenom>
+        + CanDecode<ViaCairo, StarknetAddress>,
+{
+    async fn query_cosmos_token_address_on_starknet(
+        chain: &Chain,
+        prefixed_denom: &PrefixedDenom,
+    ) -> Result<Option<StarknetAddress>, Chain::Error> {
+        let encoding = chain.encoding();
+        let ics20_contract_address = chain.query_contract_address(PhantomData).await?;
+
+        let calldata = encoding
+            .encode(prefixed_denom)
+            .map_err(Chain::raise_error)?;
+
+        let message = StarknetMessage {
+            call: Call {
+                to: ics20_contract_address.0,
+                selector: selector!("create_ibc_token"),
+                calldata,
+            },
+            counterparty_height: None,
+        };
```

Comment:
I see that you now allow returning existing token. Note that, this is a contract invoke -- so we will spend gas on it, even if we don't modify the state. Don't you think it makes sense to query via `ibc_token_address` first ? And create the token if it doesn't exist ?

Reply:
Ignore I see that, you're doing exactly this later. 👍🏼

Reply:
Yeah, here I make token creation to be no-op if a token already exists. This helps resolving the race condition where we try to create a token while a relayer is relaying and triggering the token creation at the same time.